File: server/src/main/java/com/lightbend/modelserver/store/ModelStateStoreChangeLogger.java
package com.lightbend.modelserver.store;

import org.apache.kafka.common.serialization.Serializer;
import org.apache.kafka.streams.processor.ProcessorContext;
import org.apache.kafka.streams.processor.internals.ProcessorStateManager;
import org.apache.kafka.streams.processor.internals.RecordCollector;
import org.apache.kafka.streams.state.StateSerdes;

/**
 * Created by boris on 7/11/17.
 * based on
 * https://github.com/confluentinc/examples/blob/3.2.x/kafka-streams/src/main/scala/io/confluent/examples/streams/algebird/CMSStoreChangeLogger.scala
 */
public class ModelStateStoreChangeLogger<K,V> {

    private String topic;
    private RecordCollector collector;
    private int partition;
    private StateSerdes<K, V> serialization;
    private ProcessorContext context;

    public ModelStateStoreChangeLogger(String storeName, ProcessorContext context, int partition, StateSerdes<K, V> serialization){
        topic = ProcessorStateManager.storeChangelogTopic (context.applicationId(), storeName);
        collector = ((RecordCollector.Supplier)context).recordCollector();
        this.partition = partition;
        this.serialization = serialization;
        this.context = context;
    }

    public ModelStateStoreChangeLogger(String storeName, ProcessorContext context, StateSerdes<K, V> serialization){
        this(storeName, context, context.taskId().partition, serialization);
    }

    public void logChange(K key, V value) {
        if (collector != null) {
            Serializer<K> keySerializer = serialization.keySerializer();
            Serializer<V> valueSerializer = serialization.valueSerializer();
            collector.send(this.topic, key, value, this.partition, context.timestamp(), keySerializer, valueSerializer);
        }
    }
}

File: kafkaStreamsModelServerCustomStore/src/main/scala/com/lightbend/scala/custom/store/ModelStateStoreChangeLogger.scala
package com.lightbend.scala.custom.store

import org.apache.kafka.streams.processor.ProcessorContext
import org.apache.kafka.streams.processor.internals.{ProcessorStateManager, RecordCollector}
import org.apache.kafka.streams.state.StateSerdes

class ModelStateStoreChangeLogger[K, V]
    (storeName: String, context: ProcessorContext, partition: Int, serialization: StateSerdes[K, V]){

  val topic = ProcessorStateManager.storeChangelogTopic(context.applicationId, storeName)
  val collector = context match {
    case rc: RecordCollector.Supplier => rc.recordCollector
    case _ => throw new RuntimeException(s"Expected a context that is a RecordCollector.Supplier, but got this: $context")
  }

  def this(storeName: String, context: ProcessorContext, serialization: StateSerdes[K, V]) {
    this(storeName, context, context.taskId.partition, serialization)
  }

  def logChange(key: K, value: V): Unit = {
    if (collector != null) {
      val keySerializer = serialization.keySerializer
      val valueSerializer = serialization.valueSerializer
      var ts = 0L
      try
        ts = context.timestamp
      catch {
        case t: Throwable =>
          ts = System.currentTimeMillis
      }
      collector.send(this.topic, key, value, this.partition, ts, keySerializer, valueSerializer)
    }
  }
}
